packages/sdk/src/evm/core/classes/transactions.ts

```diff
@@ -694,3 +723,129 @@ export class Transaction<TResult = TransactionResult> {
     });
   }
 }
+
+export class DeployTransaction extends TransactionContext {
+  factory: ContractFactory;
+
+  constructor(options: DeployTransactionOptions) {
+    super(options);
+    this.factory = options.factory;
+  }
+
+  encode(): string {
+    return utils.hexlify(
+      utils.concat([
+        this.factory.bytecode,
+        this.factory.interface.encodeDeploy(this.args),
+      ]),
+    );
+  }
+
+  async sign(): Promise<string> {
+    const populatedTx = await this.populateTransaction();
+    return this.signer.signTransaction(populatedTx);
+  }
+
+  async simulate() {
+    const populatedTx = await this.populateTransaction();
+    this.signer.call(populatedTx);
+  }
+
+  async estimateGasLimit(): Promise<BigNumber> {
+    try {
+      const populatedTx = await this.populateTransaction();
+      return this.signer.estimateGas(populatedTx);
+    } catch (err) {
+      // No need to do simulation here, since there can't be revert errors
+      throw this.deployError(err);
+    }
+  }
+
+  async send(): Promise<ContractTransaction> {
+    try {
+      const signedTx = await this.sign();
+      return this.provider.sendTransaction(signedTx);
+    } catch (err) {
+      throw this.deployError(err);
+    }
+  }
+
+  async execute(): Promise<string> {
+    const tx = await this.send();
+
+    try {
+      await tx.wait();
+    } catch (err) {
+      // If tx.wait() fails, it just gives us a generic "transaction failed"
+      // error. So instead, we need to call static to get an informative error message
+      await this.simulate();
+
+      // If transaction simulation (static call) doesn't throw, then throw with the message that we have
+      throw await this.deployError(err);
+    }
+
+    return utils.getContractAddress({ from: tx.from, nonce: tx.nonce });
```

Comment:
this is neat - does it parse the tx log for that address + nonce?

Reply:
It actually just does this `getAddress(hexDataSlice(keccak256(encode([ from, nonce ])), 12))` - which ig is the way to get the address for a standard direct deploy